### ISSUE TITLE
Add webp format conversion to coloractions

### ISSUE THREAD

**Issue by laryn** [backdrop-contrib/image_effects#8]
- PR based on Vincent Demers work: https://github.com/backdrop-contrib/image_effects/pull/9

 - Reference: https://www.drupal.org/project/imagecache_actions/issues/2982694#comment-13035990
 - Blocked by https://github.com/backdrop/backdrop-issues/issues/4509

**Comment by indigoxela** [comment]
Using your experimental branch I get a php warning:

`Warning: Use of undefined constant IMAGE_DERIVATIVE_TOKEN - assumed 'IMAGE_DERIVATIVE_TOKEN' (this will throw an Error in a future version of PHP) in image_effects_coloractions_image_style_url_alter() (line 861 of .../modules/image_effects/coloractions/image_effects_coloractions.module).
`
There's only this single use of IMAGE_DERIVATIVE_TOKEN, no definition for this constant anywhere, neither in your module, nor in core.

**Comment by laryn** [comment]
@indigoxela Ah, thanks. I'm guessing this is left over from the Drupal 7 issue where they added tokens on every image request. If I recall, Backdrop handled that differently and this may be completely unnecessary here.

**Comment by laryn** [comment]
Transcribing [some comments](https://github.com/backdrop/backdrop-issues/issues/4509#issuecomment-680751239) from @indigoxela here now that webp is committed in core. It's not 100% clear to me yet but will review when I get a chance (or anyone else feel free to submit a PR in the meantime...):

> Another finding: `backdrop_alter('image_style_deliver', $style, $image_uri, $derivative_uri);` is actually not necessary, because that part can already be accessed via `hook_menu_alter()` - by overriding the page callback for `$items[$directory_path . '/styles/%image_style']`.

> So the only thing his contrib module needs, is my small core change plus one additional hook - but wait, there's more...

> He needs to alter the path provided by `file_create_url()` - which already uses a hook: `hook_file_url_alter()`.

> ...very likely the only thing you need is that we add webp support to core. Everything else is (should be?) already accessible by existing hooks - at least that's what my impression is after playing with your experimental module branch.

**Comment by laryn** [comment]
I obviously haven't gotten back to this yet, but just to note that Drupal core is supporting conversion between image types, including webp as of 9.2 (June 2021):

 - https://www.lullabot.com/articles/speed-drupals-front-end-webp-images
 - https://www.drupal.org/node/3171135

**Comment by laryn** [comment]
May be a bit of overlap with this module (but I think still worth adding this image effect conversion here):

 - https://github.com/backdrop-contrib/imagewebp

**Comment by laryn** [comment]
@indigoxela I'm finally coming back to this one and am trying to follow [your advice](https://github.com/backdrop-contrib/image_effects/issues/8#issuecomment-744557824). I've duplicated and modified `image_style_deliver` and altered the page callback for `$items[$directory_path . '/styles/%image_style']`.  I've also tried to implement `hook_file_url_alter`.

On a basic level this works well, which is great!

I'm running stuck when using an image field in a View as the query parameter that this approach tacks on the URL gets double encoded (or something along those lines) and the image doesn't get loaded correctly. Any ideas?

So in summary, an image that would normally show up at this URL:

 - https://example.org/files/styles/large/public/image-filename.jpg

Successfully shows up here:

 - https://example.org/files/styles/large/public/image-filename.webp?original-extension=jpg

But when used in a Views field is rendered with this broken URL:

 - https://example.org/files/styles/large/public/image-filename.webp%3Foriginal-extension%3Djpg

**Comment by indigoxela** [comment]
> I'm running stuck when using an image field in a View as the query parameter..

I'm not sure, what that means, can you give more details? How do you "use an image field as query param"?

**Comment by laryn** [comment]
Sorry, that wasn't worded very well. I meant setting up a View that displays fields, and adding an image field in that view (which is configured to use an image style that has this convert effect active).

![Screen Shot 2022-08-25 at 11 18 18 AM](https://user-images.githubusercontent.com/871421/186717610-7a2bb2b0-abb1-462e-ae15-6c263307d999.jpg)

**Comment by indigoxela** [comment]
> ... setting up a View that displays fields

Ah, got it.

Some testing... for me your PR doesn't work at all even for "regular" image field display.

The first glitch is on the image style settings form:

![create-im-style](https://user-images.githubusercontent.com/842066/186840434-0cdfb2ca-5dcc-4570-97d6-9285d2acae32.png)

The second glitch - image field display broken:

![broken-card-image](https://user-images.githubusercontent.com/842066/186840502-7d503373-cbc6-46b5-9d9a-c622281072aa.png)

Third glitch - views:

![cards-view](https://user-images.githubusercontent.com/842066/186840550-972688ff-2fdf-4204-a531-52e7005feee9.png)

The image url param is always encoded, so all these images get "Source image not found."

I don't think this will work: `public://styles/footest_webp/public/field/image/card1-layout.webp?original-extension=png` the param is considered to be part of the path, so I end up with broken images wherever I use that image style.

**Comment by laryn** [comment]
Strange, I must have been bamboozled by a cached version when it seemed to be working part of the time.

Okay, I think I have to give up on the attempt to use existing hooks alone and have filed an issue here to allow altering `image_style_url` in case people think it would be useful to allow: https://github.com/backdrop/backdrop-issues/issues/5745

I'm also going to repurpose this issue back to it's original intent, which was to allow `webp` conversion. I'll open a follow up issue linked to the core issue above for the ability to actually change the filename after conversion.

Thanks for testing @indigoxela!

**Comment by laryn** [comment]
Follow up is here: https://github.com/backdrop-contrib/image_effects/issues/16
